File: record_android/android/src/main/java/com/llfbandit/record/methodcall/RecorderWrapper.java
package com.llfbandit.record.methodcall;

import android.app.Activity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.llfbandit.record.record.AudioRecorder;
import com.llfbandit.record.record.RecordConfig;
import com.llfbandit.record.record.stream.RecorderRecordStreamHandler;
import com.llfbandit.record.record.stream.RecorderStateStreamHandler;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodChannel;

class RecorderWrapper {
  static final String EVENTS_STATE_CHANNEL = "com.llfbandit.record/events";
  static final String EVENTS_RECORD_CHANNEL = "com.llfbandit.record/eventsRecord";

  private EventChannel eventChannel;
  private final RecorderStateStreamHandler recorderStateStreamHandler = new RecorderStateStreamHandler();
  private EventChannel eventRecordChannel;
  private final RecorderRecordStreamHandler recorderRecordStreamHandler = new RecorderRecordStreamHandler();

  @Nullable
  private AudioRecorder recorder;

  @Nullable
  private RecordConfig config;

  RecorderWrapper(@NonNull BinaryMessenger messenger) {
    eventChannel = new EventChannel(messenger, EVENTS_STATE_CHANNEL);
    eventChannel.setStreamHandler(recorderStateStreamHandler);

    eventRecordChannel = new EventChannel(messenger, EVENTS_RECORD_CHANNEL);
    eventRecordChannel.setStreamHandler(recorderRecordStreamHandler);
  }

  public void setActivity(@Nullable Activity activity) {
    recorderStateStreamHandler.setActivity(activity);
    recorderRecordStreamHandler.setActivity(activity);
  }

  void startRecordingToFile(RecordConfig config, @NonNull MethodChannel.Result result) {
    startRecording(config, result);
  }

  void startRecordingToStream(RecordConfig config, @NonNull MethodChannel.Result result) {
    startRecording(config, result);
  }

  void dispose() {
    if (recorder != null) {
      try {
        recorder.dispose();
      } catch (Exception ignored) {
      } finally {
        recorder = null;
      }
    }

    if (eventChannel != null) {
      eventChannel.setStreamHandler(null);
      eventChannel = null;
    }

    if (eventRecordChannel != null) {
      eventRecordChannel.setStreamHandler(null);
      eventRecordChannel = null;
    }
  }

  void pause(@NonNull MethodChannel.Result result) {
    if (recorder != null) {
      try {
        recorder.pause();
        result.success(null);
      } catch (Exception e) {
        result.error("-3", e.getMessage(), e.getCause());
      }
    } else {
      result.success(null);
    }
  }

  void isPaused(@NonNull MethodChannel.Result result) {
    if (recorder != null) {
      result.success(recorder.isPaused());
    } else {
      result.success(false);
    }
  }

  void isRecording(@NonNull MethodChannel.Result result) {
    if (recorder != null) {
      result.success(recorder.isRecording());
    } else {
      result.success(false);
    }
  }

  void getAmplitude(@NonNull MethodChannel.Result result) {
    if (recorder != null) {
      List<Double> amps = recorder.getAmplitude();

      Map<String, Object> amp = new HashMap<>();
      amp.put("current", amps.get(0));
      amp.put("max", amps.get(1));

      result.success(amp);
    } else {
      result.success(null);
    }
  }

  void resume(@NonNull MethodChannel.Result result) {
    if (recorder != null) {
      try {
        recorder.resume();
        result.success(null);
      } catch (Exception e) {
        result.error("-4", e.getMessage(), e.getCause());
      }
    } else {
      result.success(null);
    }
  }

  void stop(@NonNull MethodChannel.Result result) {
    if (recorder != null) {
      try {
        recorder.stop();
        result.success(config != null ? config.path : null);
      } catch (Exception e) {
        result.error("-2", e.getMessage(), e.getCause());
      }
    } else {
      result.success(null);
    }
  }

  private void startRecording(@NonNull RecordConfig config, @NonNull MethodChannel.Result result) {
    try {
      dispose();

      this.config = config;

      recorder = createRecorder(config);

      recorder.start();
      result.success(null);
    } catch (Exception e) {
      result.error("-1", e.getMessage(), e.getCause());
    }
  }

  private AudioRecorder createRecorder(@NonNull RecordConfig config) {
    return new AudioRecorder(
        config,
        recorderStateStreamHandler,
        recorderRecordStreamHandler
    );
  }
}

File: record_android/android/src/main/kotlin/com/llfbandit/record/methodcall/RecorderWrapper.kt
package com.llfbandit.record.methodcall

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.media.AudioDeviceInfo
import android.os.IBinder
import com.llfbandit.record.record.RecordConfig
import com.llfbandit.record.record.bluetooth.BluetoothReceiver
import com.llfbandit.record.record.bluetooth.BluetoothScoListener
import com.llfbandit.record.record.recorder.AudioRecorder
import com.llfbandit.record.record.recorder.IRecorder
import com.llfbandit.record.record.recorder.MediaRecorder
import com.llfbandit.record.record.stream.RecorderRecordStreamHandler
import com.llfbandit.record.record.stream.RecorderStateStreamHandler
import com.llfbandit.record.service.AudioRecordingService
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel


class RecorderWrapper(
  private val context: Context,
  recorderId: String,
  messenger: BinaryMessenger,
) : BluetoothScoListener {
  companion object {
    const val EVENTS_STATE_CHANNEL = "com.llfbandit.record/events/"
    const val EVENTS_RECORD_CHANNEL = "com.llfbandit.record/eventsRecord/"
  }

  private var eventChannel: EventChannel?
  private val recorderStateStreamHandler = RecorderStateStreamHandler()
  private var eventRecordChannel: EventChannel?
  private val recorderRecordStreamHandler = RecorderRecordStreamHandler()
  private var recorder: IRecorder? = null
  private var bluetoothReceiver: BluetoothReceiver? = null

  init {
    eventChannel = EventChannel(messenger, EVENTS_STATE_CHANNEL + recorderId)
    eventChannel?.setStreamHandler(recorderStateStreamHandler)
    eventRecordChannel = EventChannel(messenger, EVENTS_RECORD_CHANNEL + recorderId)
    eventRecordChannel?.setStreamHandler(recorderRecordStreamHandler)
  }

  fun startRecordingToFile(config: RecordConfig, result: MethodChannel.Result) {
    startRecording(config, result)
  }

  fun startRecordingToStream(config: RecordConfig, result: MethodChannel.Result) {
    if (config.useLegacy) {
      throw Exception("Cannot stream audio while using the legacy recorder")
    }
    startRecording(config, result)
  }

  fun dispose() {
    try {
      recorder?.dispose()
    } catch (_: Exception) {
    } finally {
      maybeStopBluetooth()
      stopService()
      recorder = null
    }

    eventChannel?.setStreamHandler(null)
    eventChannel = null

    eventRecordChannel?.setStreamHandler(null)
    eventRecordChannel = null
  }

  fun pause(result: MethodChannel.Result) {
    try {
      recorder?.pause()
      result.success(null)
    } catch (e: Exception) {
      result.error("record", e.message, e.cause)
    }
  }

  fun isPaused(result: MethodChannel.Result) {
    result.success(recorder?.isPaused ?: false)
  }

  fun isRecording(result: MethodChannel.Result) {
    result.success(recorder?.isRecording ?: false)
  }

  fun getAmplitude(result: MethodChannel.Result) {
    if (recorder != null) {
      val amps = recorder!!.getAmplitude()
      val amp: MutableMap<String, Any> = HashMap()
      amp["current"] = amps[0]
      amp["max"] = amps[1]
      result.success(amp)
    } else {
      result.success(null)
    }
  }

  fun resume(result: MethodChannel.Result) {
    try {
      recorder?.resume()
      result.success(null)
    } catch (e: Exception) {
      result.error("record", e.message, e.cause)
    }
  }

  fun stop(result: MethodChannel.Result) {
    try {
      if (recorder == null) {
        result.success(null)
      } else {
        recorder?.stop(fun(path) = result.success(path))
      }
    } catch (e: Exception) {
      result.error("record", e.message, e.cause)
    } finally {
      stopService()
    }
  }

  fun cancel(result: MethodChannel.Result) {
    try {
      recorder?.cancel()
      result.success(null)
    } catch (e: Exception) {
      result.error("record", e.message, e.cause)
    }

    maybeStopBluetooth()
  }

  private fun startRecording(config: RecordConfig, result: MethodChannel.Result) {
    try {
      if (recorder == null) {
        recorder = createRecorder(config)
        start(config, result)
      } else if (recorder!!.isRecording) {
        recorder!!.stop(fun(_) = start(config, result))
      } else {
        start(config, result)
      }

      startService(config)
    } catch (e: Exception) {
      result.error("record", e.message, e.cause)
    }
  }

  private fun createRecorder(config: RecordConfig): IRecorder {
    if (config.manageBluetooth) {
      maybeStartBluetooth(config)
    }

    if (config.useLegacy) {
      return MediaRecorder(context, recorderStateStreamHandler)
    }

    return AudioRecorder(
      recorderStateStreamHandler,
      recorderRecordStreamHandler,
      context
    )
  }

  private fun start(config: RecordConfig, result: MethodChannel.Result) {
    recorder!!.start(config)
    result.success(null)
  }

  ///////////////////////////////////////////////////////////
  // Service
  ///////////////////////////////////////////////////////////
//    private var mService: AudioRecordingService? = null
  private var mServiceBound = false

  private val serviceConnection = object : ServiceConnection {
    override fun onServiceConnected(className: ComponentName, service: IBinder) {
//            val binder = service as AudioRecordingService.LocalBinder
//            mService = binder.getService()
    }

    override fun onServiceDisconnected(className: ComponentName) {
//            mService = null
    }
  }

  private fun startService(config: RecordConfig) {
    if (config.service != null) {
      val intent = Intent(context, AudioRecordingService::class.java)
      intent.putExtra("title", config.service.title)
      intent.putExtra("content", config.service.content)
      context.startService(intent)

      Intent(context, AudioRecordingService::class.java).also { intent ->
        mServiceBound = context.bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
      }
    }
  }

  private fun stopService() {
    if (mServiceBound) {
      context.unbindService(serviceConnection)
      context.stopService(Intent(context, AudioRecordingService::class.java))
      mServiceBound = false
    }
  }

  ///////////////////////////////////////////////////////////
  // Bluetooth SCO
  ///////////////////////////////////////////////////////////
  private fun maybeStartBluetooth(config: RecordConfig) {
    if (config.device != null && config.device.type != AudioDeviceInfo.TYPE_BLUETOOTH_SCO) {
      maybeStopBluetooth()
      return
    }

    if (bluetoothReceiver == null) {
      bluetoothReceiver = BluetoothReceiver(context)
    }

    if (!bluetoothReceiver!!.hasListeners()) {
      bluetoothReceiver!!.register()
      bluetoothReceiver!!.addListener(this)
    }
  }

  private fun maybeStopBluetooth() {
    bluetoothReceiver?.removeListener(this)

    if (bluetoothReceiver?.hasListeners() != true) {
      bluetoothReceiver?.unregister()
    }
  }

  override fun onBlScoConnected() {
  }

  override fun onBlScoDisconnected() {
  }
}
